src/MLStyle.jl

```diff
@@ -94,13 +94,108 @@ p                end
     end
 end
 
+#=
+    help functions for `@when`
+=#
+
+@active IsMacro{s::String}(x) begin
+    @match x begin
+        Expr(:macrocall, &(Symbol("@", s)), ::LineNumberNode) => true
+        Expr(:macrocall, &(Symbol("@", s)), ::LineNumberNode, arg) => true
+        _ => false
+    end
+end
+
+@active MacroSplit{s::String}(x) begin
+    @match x begin
+        Expr(:macrocall, &(Symbol("@", s)), ::LineNumberNode, arg) => arg
+        _ => nothing
+    end
+end
+
+function RecursionMatch(block, bds=[], rets=[], default=nothing)
+    @match block begin
+        # empty block
+        Expr(:block) => (bds, rets, default)
+        
+        # skip `LineNumberNode`
+        Expr(:block, 
+            ::LineNumberNode,
+            remain...
+        ) => 
+            RecursionMatch(Expr(:block, remain...), bds, rets, default)
+        
+        # match `@when`s at middle
+        Expr(:block, 
+            MacroSplit{"when"}(bd),
+            ::LineNumberNode,
```

Comment:
Don't rely on if there's a LineNumberNode, some packages like Test.jl will remove LineNumberNode in Expr except those of `macrocall`s.